crates/astria-sequencer-relayer/src/relayer/write.rs

```diff
@@ -0,0 +1,366 @@
+use std::{
+    future::Future,
+    sync::Arc,
+    task::Poll,
+    time::Duration,
+};
+
+use astria_eyre::eyre::{
+    self,
+    WrapErr as _,
+};
+use celestia_client::{
+    celestia_types::Blob,
+    jsonrpsee::http_client::HttpClient,
+    submission::ToBlobs as _,
+};
+use futures::{
+    future::{
+        Fuse,
+        FusedFuture as _,
+    },
+    FutureExt as _,
+};
+use pin_project_lite::pin_project;
+use sequencer_client::{
+    tendermint::block::Height as SequencerHeight,
+    SequencerBlock,
+};
+use tokio::{
+    select,
+    sync::mpsc::{
+        self,
+        error::{
+            SendError,
+            TrySendError,
+        },
+        Receiver,
+        Sender,
+    },
+};
+use tokio_util::task::JoinMap;
+use tracing::{
+    debug,
+    error,
+    info,
+    instrument,
+    warn,
+    Instrument as _,
+    Span,
+};
+
+struct QueuedBlobs {
+    blobs: Vec<Blob>,
+    heights: Vec<SequencerHeight>,
+}
+
+impl QueuedBlobs {
+    fn num_blobs(&self) -> usize {
+        self.blobs.len()
+    }
+
+    fn new() -> Self {
+        Self {
+            heights: Vec::new(),
+            blobs: Vec::new(),
+        }
+    }
+
+    fn push(&mut self, mut blobs: Vec<Blob>, height: SequencerHeight) {
+        self.blobs.append(&mut blobs);
+        self.heights.push(height);
+    }
+
+    fn take(&mut self) -> TakeBlobs<'_> {
```

Comment:
Yes, I added some text to `QueuedBlobs::take`. We have this construction elsewhere in the code. The reason for having a future here (instead of something like `futures::ready(queued_blobs.take_immediately())`, where `take_immediately` is eagerly, not lazily, moving the blobs out of the queue) is async cancellation.